File: content-machine/src/lib/instagram-post-url.ts
/** Match backend `services/instagram_post_url.py` for URL-adapt / preview lookups. */

export function canonicalInstagramPostUrl(url: string): string {
  return String(url ?? "")
    .trim()
    .split("?")[0]
    .split("#")[0]
    .replace(/\/+$/, "");
}

const IG_SHORT_CODE_RE = /instagram\.com\/(?:reel|reels|p|tv)\/([^/?#]+)/i;

export function instagramPostShortCode(url: string): string {
  const m = IG_SHORT_CODE_RE.exec(String(url ?? "").trim());
  return (m?.[1] ?? "").trim();
}

export function instagramPostUrlLookupVariants(url: string): string[] {
  const seen = new Map<string, true>();
  const base = canonicalInstagramPostUrl(url);
  if (base) seen.set(base, true);
  const sc = instagramPostShortCode(url);
  if (sc) {
    for (const path of [
      `https://www.instagram.com/reel/${sc}`,
      `https://www.instagram.com/reels/${sc}`,
      `https://www.instagram.com/p/${sc}`,
      `https://www.instagram.com/tv/${sc}`,
    ]) {
      const c = canonicalInstagramPostUrl(path);
      if (c) seen.set(c, true);
    }
  }
  return [...seen.keys()];
}

export function scrapedReelRowMatchesComposerUrl(
  row: { post_url?: string | null },
  composerUrl: string,
): boolean {
  const pu = (row.post_url ?? "").trim();
  if (!pu) return false;
  const variants = new Set(instagramPostUrlLookupVariants(composerUrl));
  return variants.has(canonicalInstagramPostUrl(pu));
}

File: backend/services/instagram_post_url.py
"""Canonical Instagram post/reel URLs for UNIQUE(client_id, post_url) deduplication."""

from __future__ import annotations

import re

# Matches reel, reels, /p/, and /tv/ paths; short code is the first path segment after the type.
_IG_SHORT_CODE_RE = re.compile(
    r"instagram\.com/(?:reels|reel|p|tv)/([^/?#]+)",
    re.IGNORECASE,
)


def instagram_post_short_code(url: str) -> str:
    """Extract media short code from an Instagram reel, post, or tv URL. Empty if not parseable."""
    if not url:
        return ""
    m = _IG_SHORT_CODE_RE.search(str(url))
    return (m.group(1) or "").strip() if m else ""


def canonical_reel_url_from_short_code(short_code: str) -> str:
    """Stable /reel/{shortCode} URL for storage — same media as /p/{shortCode}."""
    sc = (short_code or "").strip()
    if not sc:
        return ""
    return canonical_instagram_post_url(f"https://www.instagram.com/reel/{sc}")


def canonical_instagram_post_url(url: str) -> str:
    """Strip whitespace, query, fragment, and trailing slash so upserts match one row per post."""
    if not url:
        return ""
    return str(url).strip().split("?")[0].split("#")[0].rstrip("/")


def instagram_post_url_lookup_variants(url: str) -> list[str]:
    """Canonical ``post_url`` strings that may exist in ``scraped_reels`` for the same IG media.

    Rows may store ``/reel/{code}`` while the user pastes ``/p/{code}`` (or the reverse).
    """
    seen: dict[str, None] = {}
    base = canonical_instagram_post_url(url)
    if base:
        seen[base] = None
    sc = instagram_post_short_code(url)
    if sc:
        for path in (
            f"https://www.instagram.com/reel/{sc}",
            f"https://www.instagram.com/reels/{sc}",
            f"https://www.instagram.com/p/{sc}",
            f"https://www.instagram.com/tv/{sc}",
        ):
            c = canonical_instagram_post_url(path)
            if c:
                seen[c] = None
    return list(seen.keys())
